src/frequenz/dispatch/_event.py

```diff
@@ -38,3 +38,57 @@ class Deleted:
 This type is used to send dispatches that were created, updated or deleted
 over the channel.
 """
+
+
+@dataclass(frozen=True)
+class Start:
+    """A request to start a dispatch.
+
+    Sent when a dispatch starts as scheduled or was
+    modified to be active after.
+
+    Possible modifications that could lead to this:
+     * Change of start_time to one now() - 5 seconds
+     * Change of active to True
+     * Change of duration to be longer
+    """
+
+    dispatch: Dispatch
+    """The dispatch that should start."""
+
+
+@dataclass(frozen=False)
+class Stop:
+    """A request to stop a running dispatch.
+
+    Sent when a dispatch naturally ends according to its duration,
+    but also when it was deleted or modified to be no longer active.
+    """
+
+    dispatch: Dispatch
+    """The dispatch that is being stopped."""
+
+
+@dataclass(frozen=True)
+class DryRunStart:
+    """A request to start a dispatch in dry run mode."""
+
+    dispatch: Dispatch
+    """The dispatch that should start in dry run mode."""
+
+
+@dataclass(frozen=True)
+class DryRunStop:
+    """A request to stop a running dispatch in dry run mode."""
+
+    dispatch: Dispatch
+    """The dispatch that is being stopped in dry run mode."""
+
+
+DispatchCommandRequest = Start | Stop | DryRunStart | DryRunStop
```

Comment:
Mmm, I don't like this approach that much, for a few reasons:

* It seems less resilient: if the dispatch actor crashes, then suddenly all active dispatches will keep running indefinitely. With the current approach we only depend on the dispatch actor to trigger the initial run, but then actors can be more independent.
* Splitting into `dry_run` and normal dispatches will probably end up in code duplication, I like better the approach of just forwarding the `dry_run` flag to the actor doing the actual work so they can simulate what the dispatch should have done as close to the real thing as possible.

Reply:
Also semantically, I think the current approach is more precise. You are notified that a dispatch is ready to be executed. It might have been started 5 minutes ago, or 1 hour ago (if it was delayed for some reason, like the LTE connection was down for a while). Start/stop commands sound more *real-time* than things might be.

What happens if a dispatch is created to run now and for one hour, while the LTE connection is down for 1.5h. Then the connection is up, the dispatcher receives the new dispatch, is for 1.5h ago and it should be done now. What do you do, do you ignore it? Do you generate a start and stop command one right after the other? All of those seem to be bad to me. With the current approach you just send a dispatch that is ready to be executed, the consumer can decide if running something that was supposed to run 1.5h ago and finish 0.5h ago makes sense or not.

I know with the commands you still get the `dispatch` object, so you could do the checks too, but still the interface seems more misleading to me.

Reply:
Well, one motivation behind having `DryRun*` like this was that it is not easy to ignore that way. Otherwise, with relying on the app developer to check the `dry_run` flag is not very safe. It can easily be overlooked or forgotten (I did in my first iteration, thus my motivation to avoid this for future devs).

Reply:
I see, that's a fair point. I agree it would be a good thing to make it harder to make mistakes, but I'm not super concerned about missing checking for `dry_run` because it is probably a very useful feature to use during development so I guess it would be hard to go unnoticed for too long. That said, we can find other mechanisms to make it harder to miss the `dry_run` flag, like making the `dry_run` attribute a property and checking if it was ever read, and if it was not read when the `Dispatch` (`ReadyDispatch`?) object is destroyed, we can log a warning or even crash (`assert` if that works in `__del__()` or just `sys.exit()` with an error). We can also add a method `disable_dry_run_check()` in case someone was to purposely ignore it (just reading the property would also be enough, but it might look uglier/more confusing).

Another approach would be to have a separate channel for dry run ready dispatches, but I guess that could lead again to code duplication, or even people now missing the existence of the dry run ready dispatches, which won't prevent the issue.

What do you think?

Reply:
I am honestly most in favor of the approach here, it makes it an non-optional required extra case in the match statement, thus being a very explicit thing the user is now forced to think about right when writing code (as opposed to relying on internal asserts or checks that would throw at runtime as you outlined as some alternatives).

A separate channel would solve it as well yes, but I don't see what we gain with an extra channel that we don't get with a "shared" channel like this PR shows.

Reply:
We had a chat and decided to mostly keep the original design because doing the split of dry run could be not very easy to use and result in more confusing user code.

We identified an important use case though, and that is notifying users about running dispatch updates, so they don't need to subscribe to the lifecycle channel, which will be very noisy. The idea is to send the dispatch again through the `ready_to_execute` channel each time it is updated. The channel might be renamed to better convey the new semantics.

@Marenz one more thing to have in mind if we are going to add a `dispatch.is_running` property is what to do with dispatches that were supposed to start and end in the past, but there some LTE interruption, and we only know about the existence of that dispatch after it was supposed to be finished. In that case checking for `is_running` won't be very useful, as it should be force, but maybe the dispatch needs to be executed anyway.

I guess this is a broader topic about scheduling. For example, IIRC, cron skips stuff that should have run in the past, but `Anacron` doesn't. [`systemd.timer`](https://www.freedesktop.org/software/systemd/man/latest/systemd.timer.html) must also deal with this, and this is what it does:

> When a system is temporarily put to sleep (i.e. system suspend or hibernation) the realtime clock does not pause. When a calendar timer elapses while the system is sleeping it will not be acted on immediately, but once the system is later resumed it will catch up and process all timers that triggered while the system was sleeping. Note that if a calendar timer elapsed more than once while the system was continuously sleeping the timer will only result in a single service activation. If WakeSystem= (see below) is enabled a calendar time event elapsing while the system is suspended will cause the system to wake up (under the condition the system's hardware supports time-triggered wake-up functionality).

For recurring dispatches there is also the question of what to do if the same dispatch was supposed to run several times, do we run it only once like systemd or do we catch up?

I guess all of this will depend on the type of dispatch, so the dispatch API should probably provide tools to deal with all these options and make documentation clear so users are aware about these possibilities and handle them in the way that makes more sense for each actor.

Reply:
> @Marenz one more thing to have in mind if we are going to add a dispatch.is_running property is what to do with dispatches that were supposed to start and end in the past, but there some LTE interruption, and we only know about the existence of that dispatch after it was supposed to be finished. In that case checking for is_running won't be very useful, as it should be force, but maybe the dispatch needs to be executed anyway.

No, we can't make guarantees like that. The app receives the dispatch when it receives it and evaluates what to do with it. 
If an app(developer) decides they must run all dispatches they have not explicitly previously run (e.g. by logging it in the actor) they can implement it themselves.

For those that care, `update_time` would be the only way to differentiate between a dispatch whose `duration` was reduced vs a dispatch who arrived past it's ending date..

Reply:
> No, we can't make guarantees like that. The app receives the dispatch when it receives it and evaluates what to do with it.
> If an app(developer) decides they must run all dispatches they have not explicitly previously run (e.g. by logging it in the actor) they can implement it themselves.
> For those that care, `update_time` would be the only way to differentiate between a dispatch whose `duration` was reduced vs a dispatch who arrived past it's ending date..

I'm not talking about updated dispatches necessarily but also created ones, suppose the LTE connection is down for 4 hours and someone created a recurrent dispatch that should run every hour for 15 minutes just after the LTE connection was down. Now 4 hours later, this library pools once more but this time there is a newly created dispatch. We will send the create event (only one) via the lifecycle channel, that's clear. Now from the user perspective, there is a dispatch that should have been executed 4 times already.

I guess you say that we just wait until the next scheduled dispatch should happen and send only that dispatch via the ready to execute channel? If this is the case, for users that need to ensure all missed scheduled dispatched would have been executed, they will have to do a lot of work.

If this library can track this down and send 4 dispatches (the same object basically) via the ready to execute channel instead, then users can more easily implement this logic, not having to schedule the missed dispatch triggers themselves, but just looking if the dispatch should be running **now** or should have ran in the past and they want to catch up (or not).

Now I'm wondering if the catch-up behavior should be actually a decision made by the actor or the user creating the dispatch, and should be part of the dispatch itself. :thinking:

If it is the later, then this library could act accordingly and send catch-up dispatches through the ready to execute channel or not automatically.

I guess this is a question for Tom, but we probably need to formulate it from scratch, I guess just pinging him here could lead to a lot of confusion if we don't present the problem properly. But first, what do you think?

Reply:
> 4 dispatches (the same object basically) via the ready to execute channel instead, then users can more easily implement this logic, not having to schedule the missed dispatch triggers themselves

I mean, I would send only one dispatch, how often it should have run can be checked locally then.
We can however provide methods in the dispatch object to make those checks easy.

Reply:
Yeah, that sounds good too. Unless we want to make the catch-up behavior part of the `Dispatch` object, in which case I think this library should send several dispatch notification if the catch-up behavior is so.